cmd/dist-receiver/main.go

```diff
@@ -0,0 +1,325 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	b64 "encoding/base64"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	config "github.com/ipfs/go-ipfs-config"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core/coreapi"
+	"github.com/ipfs/go-ipfs/core/node/libp2p"
+	"github.com/ipfs/go-ipfs/plugin/loader"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+	icore "github.com/ipfs/interface-go-ipfs-core"
+	icorepath "github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var stageNow int
+var sig chan bool
+var targets []string
+
+func main() {
+	bootstrap := flag.String("bootstrap", "", "Specify the bootstrap node")
+	cidArg := flag.String("cid", "", "Specify the CID")
+	fileName := flag.String("file", "", "Specify the file name")
+	targetDir := flag.String("dir", "", "Specify the target directory")
+
+	flag.Parse()
+
+	stageNow = 0
+	sig = make(chan bool)
+
+	server := &http.Server{
+		Addr:         "0.0.0.0:4002",
+		ReadTimeout:  5 * time.Second,
+		WriteTimeout: 5 * time.Second,
+		IdleTimeout:  600 * time.Second,
+	}
+
+	http.HandleFunc("/stage", stage)
+	http.HandleFunc("/next", next)
+	http.HandleFunc("/connect", connect)
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			panic(fmt.Errorf("failed to spawn command receiver: %s", err))
+		}
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	defer func() {
+		if err := os.RemoveAll(*fileName); err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to clean up %s: %s\n", *fileName, err)
+		}
+	}()
+
+	node, _, err := spawnEphemeral(ctx, *bootstrap)
+	if err != nil {
+		panic(fmt.Errorf("failed to spawn ephemeral node: %s", err))
+	}
+
+	if err := connectToPeers(ctx, node, []string{*bootstrap}); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to connect to root peer: %s\n", err)
+	}
+
+	stageNow = 1
+
+	<-sig
+
+	if err := connectToPeers(ctx, node, targets); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to connect to peers: %s\n", err)
+	}
+
+	cid := icorepath.New(*cidArg)
+
+	if err := node.Dht().Provide(ctx, cid); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to seed the resource file: %s\n", err)
+	}
+
+	if err := node.Pin().Add(ctx, cid); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to pin the resource file: %s\n", err)
+	}
+
+	rootNode, err := node.Unixfs().Get(ctx, cid)
+
+	if err != nil {
+		panic(fmt.Errorf("could not get file with CID: %s", err))
+	}
+
+	if err := os.RemoveAll(*fileName); err != nil {
+		panic(fmt.Errorf("could not clean previous temporary file: %s", err))
+	}
+
+	err = files.WriteTo(rootNode, *fileName)
+	if err != nil {
```

Comment:
1. Done, replaced every similar place in my code. However I do note that there are still quite some similar snippets in sealer's codebase, which I'll be not fixing now since they are not a part of my commit.